### PR TITLE
Fix gz compressing files larger than 4GB

### PR DESCRIPTION
**Pull request by thgreasi**
Change-type: patch
See: https://balena.fibery.io/Work/Project/2525

### CODE REVIEW

File: lib.coffee [hunk 1]
```diff
@@ -51,10 +51,18 @@ exports.createGzipFromParts = (parts) ->
...
-	# write length
+	CRC32_PERIOD_NUMBER = 0xFFFFFFFF # 2^32-1
+	normalizedParts = parts.map (p) ->
+		# crc32_combine(crc1, crc2, n) receives len as a 32-bit integer, so any len >= 2^32 must be reduced here.
+		if p.len <= CRC32_PERIOD_NUMBER
+			return p
+		return { crc: p.crc, len: p.len % CRC32_PERIOD_NUMBER }
+	out.append(crcUtils.crc32_combine_multi(normalizedParts).combinedCrc32)
+	# write the ISIZE length, modulo 2^32 per RFC 1952 section 2.3.1
+	# https://www.rfc-editor.org/info/rfc1952/#page-8:~:text=original%20(uncompressed)%20input-,data%20modulo%202%5E32
```
Comment (thgreasi) [inline lib.coffee:69]: image

File: lib.coffee [hunk 2]
```diff
@@ -51,10 +51,18 @@ exports.createGzipFromParts = (parts) ->
...
 	out.append(DEFLATE_END)
 	# write CRC
-	out.append(crcUtils.crc32_combine_multi(parts).combinedCrc32)
-	# write length
+	CRC32_PERIOD_NUMBER = 0xFFFFFFFF # 2^32-1
+	normalizedParts = parts.map (p) ->
+		# crc32_combine(crc1, crc2, n) receives len as a 32-bit integer, so any len >= 2^32 must be reduced here.
+		if p.len <= CRC32_PERIOD_NUMBER
+			return p
+		return { crc: p.crc, len: p.len % CRC32_PERIOD_NUMBER }
```
Comment (joshbwlng) [inline lib.coffee]: It took me some time to understand why modifying the `len` value here was actually a safe thing to do. Apparently the length itself is not stored anywhere, it is only used as a "shift count" during creation and is then thrown away. This is because CRC (Cyclic Redundancy Check) uses cyclical math and by using this modulo "trick" we can get it to the right spot without exceeding the 32-bit integer limit. Does that match your understanding @thgreasi ?

**Review by joshbwlng — APPROVED** [review]

File: lib.coffee (hunk 2 above)
Comment (thgreasi) [reply lib.coffee]: I'm still going back-and-forth on whether I fully get it or not.
I added a :robot: explanation as well for future reference and :robot: also found an edge case and fixed it.
This was the only implementation though that made the tests happy, so that's relieving.
